File: backend/app/schemas/analytics.py
"""
Schemas for Phase 4 — Behavioral Analytics.

Analytics are computed on-demand from existing food_logs, workout_logs,
and daily_reports. No separate storage needed — the source tables ARE
the behavioral record.

Analytics cover:
  - Logging consistency (did user log every day?)
  - Calorie adherence (how close to target?)
  - Workout consistency (did user work out as planned?)
  - Streak tracking
  - Trend analysis (7-day, 30-day)
  - Correction rate (how often AI estimates were wrong)
"""
from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import Field

from app.schemas.common import TrackBaseSchema


# ── Streak Info ────────────────────────────────────────────────────────────────

class StreakInfo(TrackBaseSchema):
    """Current and best logging streak."""
    current_streak_days: int
    longest_streak_days: int
    streak_started_on: date | None = None
    last_logged_date: date | None = None
    is_active_today: bool           # Has the user logged anything today?


# ── Consistency Scores ─────────────────────────────────────────────────────────

class ConsistencyBreakdown(TrackBaseSchema):
    """
    Multi-dimensional consistency analysis.
    Scores are 0.0–1.0 (higher = more consistent).
    """
    # Overall weighted composite score
    overall_score: float

    # Component scores
    logging_consistency: float      # Days with ≥1 food log / total days
    calorie_adherence: float        # Closeness to calorie target (±15% = 1.0)
    protein_adherence: float        # Closeness to protein target
    workout_consistency: float      # Workouts logged / expected workouts

    # Period (days back from today)
    period_days: int                # 7 or 30
    period_label: Literal["7d", "30d"]

    # Raw counts for transparency
    days_logged: int
    days_in_period: int
    workouts_completed: int


# ── Trend Data ─────────────────────────────────────────────────────────────────

class DailyDataPoint(TrackBaseSchema):
    """Single day's data for trend charts."""
    date: date
    calories: float
    protein_g: float | None = None
    workouts: int
    logged: bool                    # Did user log any food this day?
    consistency_score: float | None = None   # From daily_report if generated


class TrendResponse(TrackBaseSchema):
    """Time-series data for trend visualization in Flutter charts."""
    period_days: int
    data_points: list[DailyDataPoint]
    calorie_target: float | None = None
    protein_target_g: float | None = None
    average_calories: float
    average_protein_g: float | None = None


# ── Behavioral Patterns ────────────────────────────────────────────────────────

class MealAdherencePattern(TrackBaseSchema):
    """Which meals are logged most/least consistently."""
    meal_type: str
    log_frequency_pct: float        # % of days this meal was logged
    avg_calories: float
    most_common_foods: list[str]    # Top 3 foods logged for this meal


class EstimationAccuracyStats(TrackBaseSchema):
    """
    How well the AI estimation pipeline is performing.
    High correction_rate = AI needs more training data (more logs).
    """
    total_logs: int
    corrected_logs: int
    correction_rate_pct: float
    avg_calorie_delta: float        # Average absolute difference after correction
    source_breakdown: dict[str, int]   # {"memory": 45, "dataset": 30, "llm": 25}


# ── Summary Analytics ──────────────────────────────────────────────────────────

class AnalyticsSummary(TrackBaseSchema):
    """
    Complete behavioral analytics snapshot.
    Returned by GET /analytics/summary — the main analytics endpoint.
    """
    user_id: UUID

    # Streak
    streak: StreakInfo

    # Consistency (7-day and 30-day views)
    consistency_7d: ConsistencyBreakdown
    consistency_30d: ConsistencyBreakdown

    # Patterns
    meal_patterns: list[MealAdherencePattern]
    estimation_accuracy: EstimationAccuracyStats

    # Generated insights (heuristic, not AI)
    # e.g., ["You log breakfast 90% of days but skip lunch 40% of the time"]
    pattern_insights: list[str] = Field(default_factory=list)

    # Weight trend (from progress checkpoints, if any)
    checkpoints_count: int = 0
    latest_weight_kg: float | None = None
    weight_trend_kg: float | None = None    # Change since earliest checkpoint in period

    computed_at: date

File: mobile/lib/core/models/analytics.dart
class StreakInfo {
  final int currentStreakDays;
  final int longestStreakDays;
  final String? streakStartedOn;
  final String? lastLoggedDate;
  final bool isActiveToday;

  const StreakInfo({
    required this.currentStreakDays,
    required this.longestStreakDays,
    this.streakStartedOn,
    this.lastLoggedDate,
    required this.isActiveToday,
  });

  factory StreakInfo.fromJson(Map<String, dynamic> json) => StreakInfo(
        currentStreakDays: json['current_streak_days'] as int,
        longestStreakDays: json['longest_streak_days'] as int,
        streakStartedOn: json['streak_started_on'] as String?,
        lastLoggedDate: json['last_logged_date'] as String?,
        isActiveToday: json['is_active_today'] as bool,
      );
}

class ConsistencyBreakdown {
  final double overallScore;
  final double loggingConsistency;
  final double calorieAdherence;
  final double proteinAdherence;
  final double workoutConsistency;
  final int periodDays;
  final String periodLabel;
  final int daysLogged;
  final int daysInPeriod;
  final int workoutsCompleted;

  const ConsistencyBreakdown({
    required this.overallScore,
    required this.loggingConsistency,
    required this.calorieAdherence,
    required this.proteinAdherence,
    required this.workoutConsistency,
    required this.periodDays,
    required this.periodLabel,
    required this.daysLogged,
    required this.daysInPeriod,
    required this.workoutsCompleted,
  });

  factory ConsistencyBreakdown.fromJson(Map<String, dynamic> json) => ConsistencyBreakdown(
        overallScore: (json['overall_score'] as num).toDouble(),
        loggingConsistency: (json['logging_consistency'] as num).toDouble(),
        calorieAdherence: (json['calorie_adherence'] as num).toDouble(),
        proteinAdherence: (json['protein_adherence'] as num).toDouble(),
        workoutConsistency: (json['workout_consistency'] as num).toDouble(),
        periodDays: json['period_days'] as int,
        periodLabel: json['period_label'] as String,
        daysLogged: json['days_logged'] as int,
        daysInPeriod: json['days_in_period'] as int,
        workoutsCompleted: json['workouts_completed'] as int,
      );
}

class DailyDataPoint {
  final String date;
  final double calories;
  final double? proteinG;
  final int workouts;
  final bool logged;
  final double? consistencyScore;

  const DailyDataPoint({
    required this.date,
    required this.calories,
    this.proteinG,
    required this.workouts,
    required this.logged,
    this.consistencyScore,
  });

  factory DailyDataPoint.fromJson(Map<String, dynamic> json) => DailyDataPoint(
        date: json['date'] as String,
        calories: (json['calories'] as num).toDouble(),
        proteinG: (json['protein_g'] as num?)?.toDouble(),
        workouts: json['workouts'] as int,
        logged: json['logged'] as bool,
        consistencyScore: (json['consistency_score'] as num?)?.toDouble(),
      );
}

class TrendResponse {
  final int periodDays;
  final List<DailyDataPoint> dataPoints;
  final double? calorieTarget;
  final double? proteinTargetG;
  final double averageCalories;
  final double? averageProteinG;

  const TrendResponse({
    required this.periodDays,
    required this.dataPoints,
    this.calorieTarget,
    this.proteinTargetG,
    required this.averageCalories,
    this.averageProteinG,
  });

  factory TrendResponse.fromJson(Map<String, dynamic> json) => TrendResponse(
        periodDays: json['period_days'] as int,
        dataPoints: (json['data_points'] as List<dynamic>)
            .map((e) => DailyDataPoint.fromJson(e as Map<String, dynamic>))
            .toList(),
        calorieTarget: (json['calorie_target'] as num?)?.toDouble(),
        proteinTargetG: (json['protein_target_g'] as num?)?.toDouble(),
        averageCalories: (json['average_calories'] as num).toDouble(),
        averageProteinG: (json['average_protein_g'] as num?)?.toDouble(),
      );
}

class AnalyticsSummary {
  final String userId;
  final StreakInfo streak;
  final ConsistencyBreakdown consistency7d;
  final ConsistencyBreakdown consistency30d;
  final List<String> patternInsights;
  final int checkpointsCount;
  final double? latestWeightKg;
  final double? weightTrendKg;
  final DateTime computedAt;

  const AnalyticsSummary({
    required this.userId,
    required this.streak,
    required this.consistency7d,
    required this.consistency30d,
    required this.patternInsights,
    required this.checkpointsCount,
    this.latestWeightKg,
    this.weightTrendKg,
    required this.computedAt,
  });

  factory AnalyticsSummary.fromJson(Map<String, dynamic> json) => AnalyticsSummary(
        userId: json['user_id'] as String,
        streak: StreakInfo.fromJson(json['streak'] as Map<String, dynamic>),
        consistency7d: ConsistencyBreakdown.fromJson(json['consistency_7d'] as Map<String, dynamic>),
        consistency30d: ConsistencyBreakdown.fromJson(json['consistency_30d'] as Map<String, dynamic>),
        patternInsights: (json['pattern_insights'] as List<dynamic>).cast<String>(),
        checkpointsCount: json['checkpoints_count'] as int,
        latestWeightKg: (json['latest_weight_kg'] as num?)?.toDouble(),
        weightTrendKg: (json['weight_trend_kg'] as num?)?.toDouble(),
        computedAt: DateTime.parse(json['computed_at'] as String),
      );
}
